### PR TITLE
Add a Graph implementation

### PR DESCRIPTION
**Pull request by mtpereira**
This change will allow for implementing search algorithms on top of graphs.

The performance is probably sub-par and will be properly benchmarked and improved as needed.

### CODE REVIEW

File: graphs/graphs.go [hunk 1]
```diff
@@ -52,42 +52,29 @@ func New(input string) (*Graph, error) {
...
-		// Space character except newline or EOF -> ignore.
-		case unicode.IsSpace(c) && c != '\n' || c == 0:
-			continue
-		// Newline -> create a new node on next iteration.
-		case c == '\n':
-			current = nil
-		default:
+	for scanner.Scan() {
+		ids := strings.Fields(scanner.Text())
+		nodeID, err := strconv.ParseUint(ids[0], 0, 64)
```
Comment (jacoelho) [inline graphs/graphs.go]: what will happen if line is empty? ids[0]

Comment (jacoelho) [reply graphs/graphs.go]: ParseUint(ids[0], 0, 64)

although it works, you are using base 10, it isn't clear as is and relies checking implementation

ParseUint(ids[0], 10, 64)

Comment (mtpereira) [reply graphs/graphs.go]: > what will happen if line is empty? ids[0]

`scanner.Scan` returns `false` on that case and the loop isn't executed. There's a test for that case.

Comment (mtpereira) [reply graphs/graphs.go]: > ParseUint(ids[0], 0, 64)
> 
> although it works, you are using base 10, it isn't clear as is and relies checking implementation
> 
> ParseUint(ids[0], 10, 64)

Good point, it does make it clearer! Thanks!

Comment (jacoelho) [reply graphs/graphs.go]: > 
> 
> > what will happen if line is empty? ids[0]
> 
> `scanner.Scan` returns `false` on that case and the loop isn't executed. There's a test for that case.

it is fail if it contains a \n for example: https://play.golang.org/p/EXzWZystbFe

I guess it is outside project aim, more just fyi

**Pull request merged**